### PR TITLE
feat(providers): add custom headers option

### PR DESCRIPTION
**Pull request by brizental**
Relates to #75 

This still needs tests, but I thought we should discuss which tests together :)

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -23,7 +25,7 @@ rusoto_core = ">=0.32.0"
 rusoto_credential = ">=0.11.0"
 rusoto_ses = ">=0.32.0"
 rusoto_sqs = ">=0.32.0"
-sendgrid = ">=0.7.0"
+sendgrid = { git = "https://github.com/gsquire/sendgrid-rs.git" }
```
Comment (vladikoff) [inline Cargo.toml]: 0.7.1 is already available with your patch https://crates.io/crates/sendgrid/0.7.1 right?

File: src/providers/ses.rs [hunk 2]
```diff
@@ -44,44 +46,70 @@ impl Provider for SesProvider {
...
+
+        let encoded_message = encode(&format!("{}", message));
+        let mut request = SendRawEmailRequest::default();
+        request.raw_message = RawMessage {
+            data: format!("{}", encoded_message).as_bytes().to_vec(),
+        };
 
         self.client
-            .send_email(&request)
+            .send_raw_email(&request)
```
Comment (vladikoff) [inline src/providers/ses.rs]: with the amount of changes to switch to RawMessage to support headers I think we should land this PR and then spin up a different issue / PR to finalize the ALLOW_LIST passthrough for the known headers (from https://github.com/mozilla/fxa-email-service/issues/75#issuecomment-398420128 )

**Comment by vladikoff** [comment]
> This still needs tests, but I thought we should discuss which tests together :)

I wonder if we have the technology to mock out the <kbd>self.client.send</kbd> and <kbd>self.client.send_raw_message</kbd> functions to later assert that the properties of the `&request`. So we can later check the contents of the `headers` 🤔 

At a quick glance I didn't see any tests for `ses` and `sendgrid` providers in the `/providers` directory. 
image

File: Cargo.toml (hunk 1 above)
Comment (brizental) [reply Cargo.toml]: ahh yes that's right :)

File: src/providers/ses.rs (hunk 2 above)
Comment (brizental) [reply src/providers/ses.rs]: yes that's what i was thinking.

**Comment by brizental** [comment]
> At a quick glance I didn't see any tests for ses and sendgrid providers in the /providers directory.

@vladikoff there is an issue open for that ->  #21

**Comment by vladikoff** [comment]
> @vladikoff there is an issue open for that -> #21

Well here we go ❗️

File: src/providers/ses.rs [hunk 3]
```diff
@@ -44,44 +46,70 @@ impl Provider for SesProvider {
...
         cc: &[&str],
+        headers: Option<&Headers>,
         subject: &str,
         body_text: &str,
         body_html: Option<&str>,
     ) -> Result<String, ProviderError> {
-        let mut destination = Destination::default();
-        destination.to_addresses = Some(vec![to.to_string()]);
+        let mut all_headers = header::Headers::new();
+        all_headers.set(header::From(vec![self.sender.parse().unwrap()]));
```
Comment (philbooth) [inline src/providers/ses.rs]: Panicking inside a route handler is not acceptable I think. If one of these `parse` calls fails, we should fail gracefully and send an appropriate error response back to the caller.

**Comment by brizental** [comment]
@vladikoff @philbooth hey guys, I just sent in a commit for the reviews you asked for. r?

File: src/providers/ses.rs [hunk 4]
```diff
@@ -44,44 +46,70 @@ impl Provider for SesProvider {
...
-            message.body.html = Some(Content {
-                charset: None,
-                data: html.to_string(),
-            });
+        let mut message: Message<MultiPart<String>> = Message::new().with_headers(all_headers);
+
+        let mut body: MultiPart<String> = MultiPart::alternative().with_singlepart(
+            SinglePart::quoted_printable()
+                .with_header(header::ContentType(
+                    "text/plain; charset=utf8".parse().unwrap(),
```
Comment (philbooth) [inline src/providers/ses.rs]: Just noting for future reference, when we open this API up to external callers we should make it explicit in the docs that UTF-8 is assumed (or provide a param that lets them override it).

**Pull request merged**